compiler/rustc_mir_build/src/check_unsafety.rs

```diff
@@ -449,11 +451,22 @@ impl<'a, 'tcx> Visitor<'a, 'tcx> for UnsafetyVisitor<'a, 'tcx> {
                     }
                 }
             }
+            ExprKind::AddressOf { .. } => {
+                // we want to forgive one deref, so addr_of!(STATIC_MUT) works
+                self.in_addr_of = true;
```

Comment:
I think instead of making this stateful (which as you noted has the risk of accepting `addr_of!(*STATIC_MUT)`) we should make this closer to detecting `addr_of!(*ptr)` in one go as a single pattern, and treating that differently.

Reply:
Hmm, but how would I force the visitor to skip ahead over the expression I just checked so that it doesn't fire the unsafe check again on the static? Would I peer ahead for a static and then early return? We could do that, but it seemed even more likely to be erroneous than this.

Reply:
...and it seems to me we absolutely do not want to do early return whatsoever if we generalize this from applying to just `STATIC_MUT`? Because there's lots of ways to produce a pointer so `ptr` can be a wondrous variety of expressions that might themselves be `unsafe` inside.